File: canIWin/canIWin.cpp
// clang++ canIWin.cpp -std=c++11

#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    bool canIWin(int maxChoosableInteger, int desiredTotal) {
        if (maxChoosableInteger >= desiredTotal)
            return true;
        if ((maxChoosableInteger + 1) * maxChoosableInteger < 2 * desiredTotal)
            return false;
        unordered_map<int, bool> m;
        return dfs(maxChoosableInteger, desiredTotal, 0, m);
    }
    bool dfs(int& maxChoosableInteger, int desiredTotal, int used, unordered_map<int, bool>& m) {
        if (m.count(used))
            return m[used];
        for (int i = maxChoosableInteger; i >= 1; i--) {
            if ((used & (1 << i)) != 0)
                continue;
            int next_used = used | (1 << i);
            if (i >= desiredTotal || !dfs(maxChoosableInteger, desiredTotal - i, next_used, m)) 
                return m[used] = true;
        }
        return m[used] = false;
    }
};

int main()
{
    Solution s;
    cout << s.canIWin(10, 11) << endl;
    return 0;
}

File: canIWin/canIWin.java
class Solution {
    private Map<String, Boolean> memo;
    public boolean canIWin(int maxChoosableInteger, int desiredTotal) {
        if (maxChoosableInteger * (maxChoosableInteger + 1) < 2 * desiredTotal) return false;
        boolean[] visited = new boolean[maxChoosableInteger + 1];
        memo = new HashMap<>();
        return dfs(maxChoosableInteger, desiredTotal, visited);
    }
    
    public boolean dfs(int maxChoosableInteger, int remain, boolean[] visited) {
        String key = Arrays.toString(visited);
        if (memo.containsKey(key)) return memo.get(key);
        
        for (int i = maxChoosableInteger; i >= 1; i--) {
            if (visited[i]) continue;
            boolean[] op_visited = getOpVisited(visited, i);
            if (remain <= i || !dfs(maxChoosableInteger, remain - i, op_visited)) {
                memo.put(key, true);
                return true;
            }
        }
        memo.put(key, false);
        return false;
    }
    
    private boolean[] getOpVisited(boolean[] visited, int choosed) {
        boolean[] op_visited = new boolean[visited.length];
        for (int i = 0; i < visited.length; i++) {
            op_visited[i] = visited[i];
        }
        op_visited[choosed] = true;
        return op_visited;
    }
}
